proxyd/frontend_rate_limiter.go

```diff
@@ -0,0 +1,127 @@
+package proxyd
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/go-redis/redis/v8"
+)
+
+type FrontendRateLimiter interface {
+	// Take consumes a key, and a maximum number of requests
+	// per time interval. It returns a boolean denoting if
+	// the limit could be taken, or an error if a failure
+	// occurred in the backing rate limit implementation.
+	//
+	// No error will be returned if the limit could not be taken
+	// as a result of the requestor being over the limit.
+	Take(ctx context.Context, key string, max int) (bool, error)
```

Comment:
Is there's a usecase that allows a different max value for the same key? If not, we can simplify this interface with a simple `Take(ctx)` by having a mapping of keys to FrontendRateLimiters (w/ a configured max value).

Reply:
Good callout. There's no use case for this right now - fixed in next PR in the stack.